docs/tutorial.md

```diff
@@ -3,200 +3,490 @@
+  bot.chat('hi!')
+})
+```
+
+### Listening for an event
+
+The bot object has many useful [events](http://mineflayer.prismarine.js.org/#/api?id=events).
+You can listen for an event by using either `bot.on()` method or `bot.once()` method of the bot object, which takes the name of an event and a function.
+To remove specific listener you can use `bot.removeListener()` method.
+
+- `bot.on(eventName, listener)`
+  Execute the `listener` function for each time the event named `eventName` triggered.
+- `bot.once(eventName, listener)`
+  Execute the `listener` function, only once, the first time the event named `eventName` triggered.
+- `bot.removeListener(eventName, listener)`
+  Removes the specified `listener` for the event named `eventName`.
+
+Not only bot object, [`Chest`](http://mineflayer.prismarine.js.org/#/api?id=mineflayerchest), [`Furnace`](http://mineflayer.prismarine.js.org/#/api?id=mineflayerfurnace), [`Dispenser`](http://mineflayer.prismarine.js.org/#/api?id=mineflayerdispenser), [`EnchantmentTable`](http://mineflayer.prismarine.js.org/#/api?id=mineflayerenchantmenttable), [`Villager`](http://mineflayer.prismarine.js.org/#/api?id=mineflayervillager) object also have their own events!
+
+### Callbacks
+A [callback](https://en.wikipedia.org/wiki/Callback_(computer_programming)) is a function that you can give to another function, that is expected to be *called back*, generally when that function ends.  
+In Mineflayer, callbacks are often used to handle errors.
+
+```js
+bot.consume((error) => {
+  if (error) {
+    console.log(error)
+  } else {
+    console.log('Finished consuming')
   }
 })
 ```
 
+The above code will try to consume what the bot is currently holding.  
+When the consuming ends, the function that is passed along is called.  
+We can then do other things that we want to do after.  
+The function could also be called when an error occurs.
+
 ## Advanced
 
-For the following, we assume you have understood the [Basics](#basics) tutorial.
+The following concepts aren't necessary to create a Mineflayer bot, but they can be useful to understand and create more advanced bots.  
+We assume you have understood the [Basics](#basics) tutorial.
+
+### Asynchronousy
+In Javascript, asynchronousy is an important concept.  
+By default, Javascript will run everything line by line, and only go to the next line if the current line is done. This is called blocking.  
+However, sometimes you have to do something that takes a relatively long time, and you don't want your whole program to block and wait for it to finish.  
+
+Interacting with the filesystem is often done using asynchronousy, because reading and writing large files can take a long time.  
+
+```js
+const myPromise = new Promise((resolve, reject) => {
+  setTimeout( () => {
+    resolve("Success!")  // Yay! Everything went well!
+  }, 1000) 
+}) 
+
+myPromise.then((successMessage) => {
+  console.log(successMessage) 
+})
+
+myPromise.catch((failureMessage) => {
```

Comment:
it should be `error` instead of `failureMessage`

Reply:
Done